### ISSUE TITLE
Unresolvable loop in the consumer after metadata leader epoch went backwards

### ISSUE THREAD

**Issue by tymeshifter** [twmb/franz-go#119]
During consumption it seems that borkers had a problem (they are out of our control, managed service). I think it's related to [this issue](https://cwiki.apache.org/confluence/display/KAFKA/KIP-279%3A+Fix+log+divergence+between+leader+and+follower+after+fast+leader+fail+over).

This lead `franz-go` consumer to enter this kind of loop

```
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Metadata v11"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"metadata update is missing partition in topic, keeping old data"
"metadata leader epoch went backwards, ignoring update"
"metadata leader epoch went backwards, ignoring update"
"immediately updating metadata"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"read Metadata v11"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata leader epoch went backwards, ignoring update"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata leader epoch went backwards, ignoring update"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"wrote Fetch v12"
"wrote Metadata v11"
"wrote Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Metadata v11"
"metadata leader epoch went backwards, ignoring update"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata leader epoch went backwards, ignoring update"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"metadata leader epoch went backwards, ignoring update"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"immediately updating metadata"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata leader epoch went backwards, ignoring update"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"wrote Metadata v11"
"metadata leader epoch went backwards, ignoring update"
"metadata leader epoch went backwards, ignoring update"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata leader epoch went backwards, ignoring update"
"read Metadata v11"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"metadata refresh has identical topic partition data"
"immediately updating metadata"
"metadata leader epoch went backwards, ignoring update"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"read Fetch v12"
"read Fetch v12"
"wrote Metadata v11"
"read Fetch v12"
"read Fetch v12"
"read Fetch v12"
"read Fetch v12"
"read Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Metadata v11"
"metadata refresh has identical topic partition data"
"read Fetch v12"
"wrote Fetch v12"
"metadata refresh has identical topic partition data"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"immediately updating metadata"
"metadata leader epoch went backwards, ignoring update"
"metadata leader epoch went backwards, ignoring update"
"metadata update is missing partition in topic, keeping old data"
"metadata leader epoch went backwards, ignoring update"
"metadata leader epoch went backwards, ignoring update"
"metadata refresh has identical topic partition data"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Metadata v11"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"metadata refresh has identical topic partition data"
"metadata refresh has identical topic partition data"
"immediately updating metadata"
"metadata refresh has identical topic partition data"
"metadata leader epoch went backwards, ignoring update"
"metadata leader epoch went backwards, ignoring update"
"metadata leader epoch went backwards, ignoring update"
"read Metadata v11"
"metadata leader epoch went backwards, ignoring update"
"metadata refresh has identical topic partition data"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Metadata v11"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
"read Metadata v11"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"metadata leader epoch went backwards, ignoring update"
"metadata leader epoch went backwards, ignoring update"
"metadata leader epoch went backwards, ignoring update"
"metadata leader epoch went backwards, ignoring update"
"metadata refresh has identical topic partition data"
"immediately updating metadata"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"metadata refresh has identical topic partition data"
"metadata update is missing partition in topic, keeping old data"
"metadata update is missing partition in topic, keeping old data"
"wrote Metadata v11"
"wrote Fetch v12"
"read Fetch v12"
"wrote Fetch v12"
```

I don't know where this thing starts or ends but it's looping and it seems it can't recover on it's own after this kind of fault. Please advise if this is a bug or not in the library or if it can be managed more gracefully.

Restarting process resolved the issue and upon reconnection operation continued without issues.

**Comment by twmb** [comment]
If I'm understanding this correctly, you may have unclean leader election enabled, and one of the brokers became leader and then died before the others agreed on this new leader? If so, it's possible the client requested metadata from that higher-epoch leader before it died, and the rest of the cluster has a lower epoch.

Internally, the client does not allow "rewinding" the metadata _if_ the broker metadata response includes leader epochs (this is newer from 2.1+): an older epoch could mean the broker is returning out of date information, and we don't want to trust that. However, in this case, the older epoch actually is the current reality.

There's not much to do within the client for this, I could add an option to allow epoch rewinds. This may cause a bit of partition flapping in the client when partitions move around on brokers, but ultimately isn't the most harmful. I'm not sure I'd want to remove the internal check entirely.

If this is related to KIP-279, then the fix they propose was introduced in 2.0. Since the metadata response only includes leader epoch's in 2.1+, I'm not sure this is due to KIP-279. What is the remote Kafka version? Is it custom, and if so, does it support OffsetForLeaderEpoch v1+?

---

Unrelated, but did you delete a topic & recreate it with fewer partitions? Those `missing partition in topic` logs should only happen in one of two cases:
- the broker is out of date / split from the cluster & does not know of newly added partitions to an existing topic
- a topic was deleted and recreated with fewer partitions
If a topic was recreated, the client does not know how to forget about the now-missing partitions (this is to err on the side of safety in case the problem is due to the former case).

Also, what version of the client are you running?

**Comment by tymeshifter** [comment]
> Also, what version of the client are you running?

Client version 1.2.2

> Unrelated, but did you delete a topic & recreate it with fewer partitions?

No, nothing of sort. This is managed Kafka solution so broker administration is out our hands. The highest probability is that broker failover happened with the scenario you described, where new broker is reporting invalid metadata (epoch in the past).

This is most probably not a bug in the client implementation of the protocol. Client entered invalid protocol state due to issues with the broker and hard restart solved the issue. But from the perspective of someone who uses the client there was no possibility to know that this is even happening because client is just locked in its internal loop unable to resolve the issue on its own and operator has no feedback about what is going on (besides debug log output which is noisy and turned off by default).

So my issue is from engineering (API) perspective where client should represent a blackbox with input, processing, and output, but in this case it only provides input.

Or maybe it is a bug if this will happen every time when there is failover on the broker side and it will enter metadata looping unable to be used from that point on. So from my perspective, client should either (somehow) return an error when unresolvable state is reached or, find way to get out of this state on its own.

**Comment by twmb** [comment]
I can think of a few options, and I think the cleanest one would be to go with the one I thought of today (so, glad I waited):

- Add a new hook that users can use when the epoch goes backwards. Something like, `OnLeaderEpochRewind` or `OnMetadataLeaderEpochRewind`. I'm not so much a fan of this option because it punts the responsibility of a confusing situation to users. Sometimes it's ok to have a temporary blip of going backwards, it's not ok for the blip to become permanent.

- Add a callback client option that provides basically the same thing. This has an additional downside of me not being able to plug in to the standard metric plugins that I already provide: even more responsibility is punted to the user.

- Allow epoch rewinds if the epoch is continually stale for a (potentially) configurable amount of retries, defaulting to maybe 10.

I prefer the third option and will likely push that (tonight?). It's not entirely detrimental for the epoch to go backwards, because it's also used & validated when consuming, and in the case of remapping to the incorrect broker, the broker can just reply "talk to somebody else". If consuming sees `FECNED_LEADER_EPOCH`, the client (currently) notifies the user if data loss is detected in an injected fetch.

So, the current plan:
- track an epoch rewind
- if a rewind is noticed 5 times in a row, accept the rewind and log

**Comment by twmb** [comment]
This has been fixed in v1.3.0
